keras_nlp/models/gpt2/gpt2_causal_lm.py

```diff
@@ -0,0 +1,333 @@
+
+import tensorflow as tf
+from tensorflow import keras
+
+import keras_nlp
+from keras_nlp.models.gpt2.gpt2_backbone import GPT2Backbone
+from keras_nlp.models.gpt2.gpt2_causal_lm_preprocessor import (
+    GPT2CausalLMPreprocessor,
+)
+from keras_nlp.models.gpt2.gpt2_presets import backbone_presets
+from keras_nlp.utils.pipeline_model import PipelineModel
+from keras_nlp.utils.python_utils import classproperty
+
+
+@keras.utils.register_keras_serializable(package="keras_nlp")
+class EmbeddingMapping(keras.layers.Layer):
+    """A layer multiplying model outputs by the token embedding.
+
+    This layer is used to map model outputs to logits over all vocab tokens.
+    It's used in `GPT2CausalLM` to calculate next token's probability.
+
+    Args:
+        embedding_layer: a `tf.keras.layers.Embedding` instance, the token
+            embedding layer.
+    """
+
+    def __init__(self, embedding_layer, name="embedding_mapping", **kwargs):
+        super().__init__(name=name, **kwargs)
+        self.embedding_layer = embedding_layer
+
+    def call(self, inputs):
+        return tf.matmul(
+            inputs,
+            self.embedding_layer.embeddings,
+            transpose_b=True,
+        )
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
+            {
+                "embedding_layer": keras.layers.serialize(self.embedding_layer),
+            }
+        )
+        return config
+
+    @classmethod
+    def from_config(cls, config):
+        config["embedding_layer"] = keras.layers.deserialize(
+            config["embedding_layer"],
+        )
+        return cls(**config)
+
+
+@keras.utils.register_keras_serializable(package="keras_nlp")
+class GPT2CausalLM(PipelineModel):
+    """GPT2 Causal LM task model.
+
+    Causal LM is predicting the next token based on previous tokens, which is
+    the way GPT2 gets pretrained. Users can finetune `GPT2CausalLM` to generate
+    text similar to the custom dataset. `GPT2CausalLM` also has a public method
+    `generate()`, which generates text based on given prompt.
+
+    This model can optionally be configured with a `preprocessor` layer, in
+    which case it will automatically apply preprocessing to raw inputs during
+    `fit()`, `predict()`, and `evaluate()`. This is done by default when
+    creating the model with `from_preset()`.
+
+    Disclaimer: Pre-trained models are provided on an "as is" basis, without
+    warranties or conditions of any kind. The underlying model is provided by a
+    third party and subject to a separate license, available
+    [here](https://github.com/openai/gpt-2).
+
+    Args:
+        backbone: A `keras_nlp.models.GPT2Backbone` instance.
+        preprocessor: A `keras_nlp.models.GPT2CausalLMPreprocessor` or `None`.
+            If `None`, this model will not apply preprocessing, and inputs
+            should be preprocessed before calling the model.
+
+    Examples:
+
+    Example usage.
+    ```python
```

Comment:
In keeping with recent changes to our pretrained models generally, let's do a few changes here...

 - Show `from_preset` examples at the class level docstring.
 - Start with the simple workflow first (e.g. from_preset with included preprocessing), and progressively get more complex as we give further examples.

Reply:
Also: should we save the text generation examples for the `generate` docstring, which is currently very short? Maybe you're worried no one will look there?

Reply:
Yea right, i feel that people won't bother reading the docstring of each method, so putting it with the docstring increases the visibility.

Re the finetuning stuff, it's something we need to figure out. Using top-k on a pretrained GPT2 does provide some decent generations in my recent experiments, I will experiment more before adding the statement "without finetuning the generated text is of low quality", which sounds pretty strong.

Reply:
I do somewhat feel that the class level docstring should be a good introduction to the main use cases of class as a whole (to the extent they fit in docstrings at all). If `generate()` is the main feature of the class, we should show `generate()` front and center here.

This fits with a pattern we are developing with `from_preset` (the clsmethod docstring is now minimal, the class docstring will show more complete examples), and is something I am also used to seeing in core Keras (for example with [adapt](https://www.tensorflow.org/api_docs/python/tf/keras/layers/StringLookup?version=nightly) on StringLookup).

Overall this feels like a consistent pattern we can stick to. And has the advantage of putting our most useful code examples (whatever they may be) near the top of our rendered documentation pages.

Reply:
OK, no worries. Maybe we can leave in note in the `generate` docstring to look for examples in the class docstring? Just an idea